Make updates for molecule to test tripleo_collectd

elfiesmelfie opened the pull request:
Added the defaults to this role to mimic THT defaults, so that this can be tested separately to THT, in a shorter test cycle.

elfiesmelfie committed:
Make updates for tripleo to work

elfiesmelfie committed:
[molecule] Add initial default scenario

This adds the default molecule test, that runs the role with default
config options.

A reusable Verify play is included, which will run
* basic checks on the created configs (defined in common/verify.yml)
* stf functional tests against the collectd-test container (common/verify.yml)
* scenario-specific checks defined in scenario_name/verify.yml

common/verify kills existing collectd instances before starting the
service again. This means that running converge and verify multiple
times gives consistent results.

The defaults/main.yml was generated by the
``tripleo-ansible-conversion-starter`` and uses the same defaults as
THT to make testing possible without TripleO being used.
defaults/main.yml provides the values that would normally be passed
through from THT. This also helps maintain consistent behaviour in
case any THT params are removed.

Separate collectd_config task from deploy_steps_tasks.yml,
this makes it easier to test just the collectd
configuration tasks having to call throught the a whole
deploy stage (which has other consequences that don't
necessarily need testing when checking that tripleo_collectd
generates usable configs).

elfiesmelfie commented:
Eventually, I think the way the tests are run will have to change, but for now, we have some tests that are ensuring that tripleo_collectd produces a valid collectd config.

Currently, these tests are running the ansible role IN the created container, then running checks in the container, and then running stf-functional tests from the host.

The next iteration may re-shuffle the test stages, and create sets of configs on the host, then launch the container, then run checks in the container, and then run stf-functional test from the host.

csibbitt commented on defaults/main.yml:
```diff
@@ -0,0 +1,333 @@
+---
+###############################################################################
+# TODO: Add the default values in here for any of the converted heat params
```
Is this TODO done now, or are there other heat params that need to be added?

elfiesmelfie replied:
This has been removed

csibbitt commented on molecule/default/molecule.yml:
```diff
@@ -0,0 +1,37 @@
+    - lint
+    - destroy
+    - dependency
+    # - syntax
```
I assume this is commented out because the ansible-lint basically takes care of it?

elfiesmelfie replied:
Yes

csibbitt commented on molecule/default/molecule.yml:
```diff
@@ -0,0 +1,37 @@
+    - create
+    - prepare
+    - converge
+    # - idempotence
```
Why is this one commented out?

elfiesmelfie replied:
The config dir is wiped before creating the configs, so that there aren't any extra plugins left that weren't in the plugin list. This means that things are changed every time, so the idempotence check will fail

csibbitt commented on molecule/common/prepare.yml:
```diff
@@ -0,0 +1,13 @@
+    - name: "Add config.json"
+      copy:
+        src: "{{ playbook_dir }}/../../tests/kolla_config_files/config.json"
+        dest: /var/lib/kolla/config_files/config.json
```
I assume this is the #208 violation? Any reason you don't just add the required perms?

elfiesmelfie replied:
This is possibly going away. I'm considering whether it's a good idea to run the deploy_steps_tasks and create the configs on the host, then mount them into the container. However, there are some side effects to running the tests, namely that there's some additional setup tasks taking place in [deploy_steps_tasks.yml ](https://github.com/infrawatch/tripleo-collectd-ansible-role/blob/molecule_tests/tasks/deploy_steps_tasks.yml)(which is called from tripleo_collectd/tasks/main.yml, using the deploy_stage var), and this is what is run by molecule.

elfiesmelfie replied:
I am planning on separating the "configure_collectd" task out of deploy_steps_tasks.yml, so that it can be moved to a different stage later, if required (it'll also make it easier to test this action on its own instead of within a deploy step with other actions run as part of the same step.

csibbitt commented on molecule/default/requirements.yml:
```diff
@@ -0,0 +1,9 @@
+---
+roles:
+  - src: git+https://github.com/infrawatch/functional-tests
+    name: functional-tests
```
Is this the #106 violation? AFAICT nothing in the code actually requires the `-`. Would changing this to `functional_tests` (here and L54 of verify.yml) allow us to remove the exception from skip_list?

elfiesmelfie replied:
That might have caused it too, but I thought it was because the ansible roles patch was set to ../ in the ansible.cfg that is created by .travis.yml, and the repo is called tripleo-collectd-ansible-role.

I can update this and check

elfiesmelfie replied:
Seems like it was functional tests, and not tripleo-collectd-ansible-role itself that was the issue, as I originally assumed.

csibbitt replied:
Looks like it was both? Latest Travis output has `[106] Role name tripleo-collectd-ansible-role does not match ``^[a-z][a-z0-9_]+$`` pattern`.

csibbitt reviewed (approved):
I think I'll need to spend some more time with main.yml to understand the variable mapping and lookups issues you're dealing with, but overall this is a good step forward!

Only objections have to do with possibly unnecessary ansible-lint exclusions, but I'm not too concerned about them.

elfiesmelfie commented on defaults/main.yml:
```diff
@@ -0,0 +1,333 @@
+
+# Corresponds to RoleName in THT
+# Role name on which the service is applied
+role_name: 
```
Why is this valid? Is it reading it as an empty string?

This should be updated, I guess.

elfiesmelfie committed:
[amqp][molecule] Add tests for amqp connection configurations

Add test for:
* amqp_default_interval
* amqp_connection

elfiesmelfie committed:
[collectd_connection] Update collectd_connection enablement

Also adds molecule tests for connection_type=network

elfiesmelfie committed:
Update how plugins are enabled

Previously, the plugins were enabled by appending to the
collectd_extra_plugins list in tasks/main.yml. This is not
the intended usage of collectd_extra_plugins, which is
passed through from TripleO.
Now, the collectd_plugins var (to be passed to
collectd_config) is created at the beginning of
tasks/main.yml and updated by:
* combining the collectd_default_plugins and
  collectd_extra_plugins from TripleO.
* adding each additional plugin as different parameters are
  parsed.

elfiesmelfie committed:
[logfile] Enable the logfile plugin

elfiesmelfie committed:
Add in hardcoded config values from THT

These corresponds to the puppet config vars that were set
in the outputs section [1]

Logfile enablement is included in a separate change to
tasks/main.yml

[1] https://github.com/openstack/tripleo-heat-templates/blob/6773fd/deployment/metrics/collectd-container-puppet.yaml#L416-L453

csibbitt commented on .ansible-lint:
```diff
@@ -1,6 +1,11 @@
    - '602'  # Don't compare to empty string
    - '701'  # Role info should contain platforms
+warn_list:
+   - '502'  # All tasks should be named
```
I know this is fairly minor, but it gives me pause to see this list growing and growing. 

* 106 I understand because it conflicts with our other conventions.
* 208 Is suspect - it should be easy to just specify the permissions.
* 502 is the same - why not follow the convention and name the tasks?
* 301 and 306 are a bit concerning, maybe I'll see them as a continue down the review

elfiesmelfie replied:
208 is a TODO
502 is also TODO, I need to go back and name all tasks at some point, there are quite a few blocks or debugs.
301 is there because it is flagged every time shell or command is used to get some information ( I think I have to add a changed_when option)
306 is new because it recently started getting flagged.

csibbitt commented on molecule/amqp_default_interval/converge.yml:
```diff
@@ -0,0 +1,13 @@
+        step: 1
+        # This is not the right things to put here... need to check tripleo_collectd
+        collectd_connection_type: 'amqp1'
+        collectd_amqp_interval: -666
```
What is this testing? The rejection of an invalid interval config?

elfiesmelfie replied:
Checking whether the default interval is set.
If CollectdAmqpInterval is set to -666, the collectd Interval is used, rather than a plugin-specific interval.
The original behaviour is in the two code snippits below:
 https://github.com/openstack/tripleo-heat-templates/blob/master/deployment/metrics/collectd-container-puppet.yaml#L383-L384
https://github.com/openstack/tripleo-heat-templates/blob/master/deployment/metrics/collectd-container-puppet.yaml#L484-L488

csibbitt commented on molecule/amqp_default_interval/verify.yml:
```diff
@@ -0,0 +1,18 @@
+        grep "Interval" {{ collectd_conf_output_dir }}
+      register: interval
+      failed_when:
+        - "{{ interval.stdout | length != 0 }}"
```
:-1: I don't think this test works. I think that `grep "Interval" /etc/collectd.d/` will always return a length 0 stdout, and the following on stderr: `grep: /etc/collectd.d/: Is a directory`. I think you mean to add a `-R` to your grep?

IIUC this test is attempting to verify that no "Interval" directive appears if we've specified an invalid value for collectd_amqp_interval?

elfiesmelfie replied:
Oh wow, this is a big boo boo. 
It should be checking for Interval in {{ collectd_conf_output_dir }}/amqp1.conf, to make sure that the global Interval is not being overridden by a plugin-specific Interval, when the CollectdAmqpInterval is set to -666 (i.e. default).

(See above for links)

csibbitt commented on tasks/main.yml:
```diff
@@ -1,83 +1,132 @@
 - name: "Update plugin list"
   set_fact:
-    collectd_extra_plugins: "{{ (collectd_extra_plugins + 'cpu df load connectivity intel_rdt ipmi procevent'.split() ) | unique }}"
+    collectd_plugins: "{{ (collectd_plugins + 'cpu df load connectivity intel_rdt ipmi procevent'.split() ) | unique }}"
```
I couldn't find a copy of the THT's you're testing against, but if `collectd_plugins` is being passed in from Heat, I think you'll find that this code does not work. See here for details: https://gist.github.com/csibbitt/95db8af1ca882a9d7bee4ef09d53300d

Same for all of the other variables below. I'm going to be experimenting with this more during the week, but my current understanding is that if a variable is passed in using the method we've identified previously, you are unable to mutate them with set_fact.

elfiesmelfie replied:
THT has the `CollectdDefaultPlugins` and `CollectdExtraPlugins` parameters [1], which translate to `collectd_default_plugins` and `collectd_extra_plugins`, respectively. 

collectd_config uses the `collectd_plugins` var, which is what's created in L10-13 in this file, by combining the parameters from THT (`collectd_default_plugins` and `collectd_extra_plugins`, as they're renamed in ansible) (as was previously done by THT in [2]).

As different plugins are enabled, the collectd_plugins var (created in this file) is extended, and is fed to collectd_config at the end.
[1] https://github.com/openstack/tripleo-heat-templates/blob/master/deployment/metrics/collectd-container-puppet.yaml#L79-L107
[2] https://github.com/openstack/tripleo-heat-templates/blob/master/deployment/metrics/collectd-container-puppet.yaml#L454-L472

csibbitt replied:
Okay, that looks like it will work. I see now where collectd_plugins comes from and how you've dealt with this problem. I'll try to emulate it.

csibbitt reviewed (commented):
A few potential problems, but definitely getting there! Thanks for the good work.

elfiesmelfie committed:
[amqp_default_interval] Fix check for Interval

csibbitt commented on molecule/amqp_default_interval/verify.yml:
```diff
@@ -12,7 +12,7 @@
       register: interval
       failed_when:
-        - "{{ interval.stdout | length != 0 }}"
+        - "{{ ( interval.stdout | length != 0 ) or ( interval.stderr_lines | length != 0 ) }}"
```
:+1:

elfiesmelfie committed:
[travis] Update molecule versions

A newer version of molecule was being installed and causing
failures in the gate due to permissions on the containers.
This change pins the molecule version so that this issue
isn't hit
